### PR TITLE
[SHELL32] RecycleBin5: Make it C++

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

Modernize code.
JIRA issue: [CORE-19595](https://jira.reactos.org/browse/CORE-19595)

## Proposed changes

- Rewrite `RecycleBin5` in C++.

### CODE REVIEW

File: dll/win32/shell32/shellrecyclebin/recyclebin_v5.cpp [hunk 1]
```diff
@@ -91,94 +91,98 @@ IntDeleteRecursive(
...
-
-    LPWSTR VolumePath;
-    WCHAR Folder[ANY_SIZE]; /* [drive]:\[RECYCLE_BIN_DIRECTORY]\{SID} */
+public:
+    RecycleBin5(
+        LPCWSTR VolumePath,
+        LPCWSTR RecycleBinDirectory,
+        LPWSTR StringSid,
+        PTOKEN_USER TokenUserInfo,
+        LPWSTR pszFolder);
```
Comment (HBelusca) [inline dll/win32/shell32/shellrecyclebin/recyclebin_v5.cpp:102]: SAL2 annotations perhaps?

File: dll/win32/shell32/shellrecyclebin/recyclebin_v5.cpp [hunk 2]
```diff
@@ -242,26 +246,26 @@ RecycleBin5_RecycleBin5_DeleteFile(
...
+    pHeader = (PINFO2_HEADER)MapViewOfFile(m_hInfoMapped, FILE_MAP_WRITE, 0, 0, 0);
     if (!pHeader)
     {
         hr = HRESULT_FROM_WIN32(GetLastError());
         goto cleanup;
     }
 
     /* Get number of entries */
-    FileSize.u.LowPart = GetFileSize(s->hInfo, &FileSize.u.HighPart);
+    FileSize.u.LowPart = GetFileSize(m_hInfo, &FileSize.u.HighPart);
```
Comment (HBelusca) [inline dll/win32/shell32/shellrecyclebin/recyclebin_v5.cpp:268]: Note: Below (line 278+) there's that GetFileSizeEx call similar to the one I mentioned in your other PR.

File: dll/win32/shell32/shellrecyclebin/recyclebin_v5.cpp [hunk 3]
```diff
@@ -305,10 +309,10 @@ RecycleBin5_RecycleBin5_DeleteFile(
         pDeletedFile->dwRecordUniqueId = pLastDeleted->dwRecordUniqueId + 1;
     }
     pDeletedFile->dwDriveNumber = tolower(szFullName[0]) - 'a';
-    _snwprintf(DeletedFileName, MAX_PATH, L"%s\\D%c%lu%s", s->Folder, pDeletedFile->dwDriveNumber + 'a', pDeletedFile->dwRecordUniqueId, Extension);
+    _snwprintf(DeletedFileName, MAX_PATH, L"%s\\D%c%lu%s", m_Folder, pDeletedFile->dwDriveNumber + 'a', pDeletedFile->dwRecordUniqueId, Extension);
```
Comment (HBelusca) [inline dll/win32/shell32/shellrecyclebin/recyclebin_v5.cpp:312]:
```suggestion
    _snwprintf(DeletedFileName, _countof(DeletedFileName), L"%s\\D%c%lu%s", m_Folder, pDeletedFile->dwDriveNumber + 'a', pDeletedFile->dwRecordUniqueId, Extension);
```

File: dll/win32/shell32/shellrecyclebin/recyclebin_v5.cpp [hunk 4]
```diff
@@ -672,31 +645,80 @@ RecycleBin5_Create(
...
+        m_hr = HRESULT_FROM_WIN32(GetLastError());
+        return;
+    }
+    *p = UNICODE_NULL;
+    m_VolumePath = p + 1;
+    wcscpy(m_VolumePath, VolumePath);
+}
+
+EXTERN_C
 HRESULT RecycleBin5_Constructor(IN LPCWSTR VolumePath, OUT IUnknown **ppUnknown)
```
Comment (HBelusca) [inline dll/win32/shell32/shellrecyclebin/recyclebin_v5.cpp:705]: Similarly to the RecycleBin5Enum PR, I wonder why not more of the code contained here isn't in the actual C++ constructor of the RecycleBin5 class.

Comment (katahiromz) [reply dll/win32/shell32/shellrecyclebin/recyclebin_v5.cpp]: Used `Init()`.

File: dll/win32/shell32/shellrecyclebin/recyclebin_v5.cpp (hunk 1 above)
Comment (katahiromz) [reply dll/win32/shell32/shellrecyclebin/recyclebin_v5.cpp]: OK.

**Pull request merged**